src/System.Linq.Queryable/src/System/Linq/CachedReflection.cs

```diff
@@ -0,0 +1,862 @@
+              .MakeGenericMethod(TSource, TKey);
+
+        private static MethodInfo s_Union_TSource_2;
+
+        public static MethodInfo Union_TSource_2(Type TSource) =>
+             (s_Union_TSource_2 ??
+             (s_Union_TSource_2 = GetMethodInfoOf<IQueryable<object>, IEnumerable<object>, IQueryable<object>>(Queryable.Union<object>).GetGenericMethodDefinition()))
+              .MakeGenericMethod(TSource);
+
+        private static MethodInfo s_Union_TSource_3;
+
+        public static MethodInfo Union_TSource_3(Type TSource) =>
+             (s_Union_TSource_3 ??
+             (s_Union_TSource_3 = GetMethodInfoOf<IQueryable<object>, IEnumerable<object>, IEqualityComparer<object>, IQueryable<object>>(Queryable.Union<object>).GetGenericMethodDefinition()))
+              .MakeGenericMethod(TSource);
+
+        private static MethodInfo s_Where_TSource_2;
+
+        public static MethodInfo Where_TSource_2(Type TSource) =>
+             (s_Where_TSource_2 ??
+             (s_Where_TSource_2 = GetMethodInfoOf<IQueryable<object>, Expression<Func<object, bool>>, IQueryable<object>>(Queryable.Where<object>).GetGenericMethodDefinition()))
+              .MakeGenericMethod(TSource);
+
+        private static MethodInfo s_Where_Index_TSource_2;
+
+        public static MethodInfo Where_Index_TSource_2(Type TSource) =>
+             (s_Where_Index_TSource_2 ??
+             (s_Where_Index_TSource_2 = GetMethodInfoOf<IQueryable<object>, Expression<Func<object, int, bool>>, IQueryable<object>>(Queryable.Where<object>).GetGenericMethodDefinition()))
+              .MakeGenericMethod(TSource);
+
+        private static MethodInfo s_Zip_TFirst_TSecond_TResult_3;
+
+        public static MethodInfo Zip_TFirst_TSecond_TResult_3(Type TFirst, Type TSecond, Type TResult) =>
+             (s_Zip_TFirst_TSecond_TResult_3 ??
+             (s_Zip_TFirst_TSecond_TResult_3 = GetMethodInfoOf<IQueryable<object>, IEnumerable<object>, Expression<Func<object, object, object>>, IQueryable<object>>(Queryable.Zip<object, object, object>).GetGenericMethodDefinition()))
+              .MakeGenericMethod(TFirst, TSecond, TResult);
+
+        private static MethodInfo GetMethodInfoOf<T1, R>(Func<T1, R> f) => GetMethodInfoOf((Delegate)f);
+        private static MethodInfo GetMethodInfoOf<T1, T2, R>(Func<T1, T2, R> f) => GetMethodInfoOf((Delegate)f);
+        private static MethodInfo GetMethodInfoOf<T1, T2, T3, R>(Func<T1, T2, T3, R> f) => GetMethodInfoOf((Delegate)f);
+        private static MethodInfo GetMethodInfoOf<T1, T2, T3, T4, R>(Func<T1, T2, T3, T4, R> f) => GetMethodInfoOf((Delegate)f);
+        private static MethodInfo GetMethodInfoOf<T1, T2, T3, T4, T5, R>(Func<T1, T2, T3, T4, T5, R> f) => GetMethodInfoOf((Delegate)f);
+        private static MethodInfo GetMethodInfoOf<T1, T2, T3, T4, T5, T6, R>(Func<T1, T2, T3, T4, T5, T6, R> f) => GetMethodInfoOf((Delegate)f);
+        private static MethodInfo GetMethodInfoOf(Delegate f) => f.GetMethodInfo();
```

Comment:
Could get rid of this helper method. My T4 file supports generating two approaches, one that uses method group conversion, and one that uses expression trees. For the latter, this method is parameterized by `LambdaExpression` and gets called by the other overloads. If we decide to go with the method group conversion (I don't think there are caveats to doing that over expression tree construction; both result in `ldftn` for the same method), this can be simplified to remove this level of indirection (though my guess would be that this overload would get inlined by the JIT).